### DISCUSSION TITLE
Writing derivative sync plugins

### DISCUSSION THREAD

**Discussion by p-zach** [bmeares/Meerschaum#59] [Q&A] [answered]
I've read [this answer](https://github.com/bmeares/Meerschaum/discussions/56#discussioncomment-2349747) about how to make changes to active plugins. Could I see a more verbose solution on how to "make a simple collection pipe to get the raw data, then create derivative pipes to transform the raw data"? I've got the fetch plugin running, but I'm unclear as to how exactly sync plugins work/how to write them even after watching [LM#4](https://www.youtube.com/watch?v=t9tFD4afSD4). Also, does the sync function exist in the same .py file as the fetch function? If not, are they still separate plugins even if one uses data from the other? Thanks.

**Comment by bmeares** [comment, marked as answer]
Hi @p-zach, thanks for posting. Plugins are regular Python modules, so you can either use a single .py file or make your plugin a directory with an `__init__.py` in case you need submodules. The `sync()` function overrides `fetch()` as the entry point into your plugin. Here's a list of the [special functions](https://meerschaum.io/reference/plugins/writing-plugins/).

Anyway, the derivative pipes pattern I mentioned (10:00 in LM#4) is one way to cluster together your related pipes. You could chain together many different plugins but that can get messy quickly (however I'm working on a dependency graph for pipes in v0.6.0 which should make defining relationships between plugins easier).

One useful trick is to write logic on the pipes' keys; that is, only fetch data from your remote source (i.e. the Twitter API) if `pipe.location_key` is `None`. Then you can decide how to handle what the derivatives look like based on their metrics or locations. That way, they're all linked by the same key (the plugin name) which makes management easier.

Here's a quick snippet to illustrate. I'm assuming you have a function called `fetch()` that returns a dataframe. I'm on mobile so please forgive any formatting issues:

```python
import meerschaum as mrsm

def sync(pipe, **kw):
    """
    Option A: Explicitly define derivatives in code.
    """
    if pipe.location_key:
        return
    pipe.sync(fetch(pipe, **kw), **kw)
    child_pipe = mrsm.Pipe(
        pipe.connector_keys,  # e.g. 'plugin:twitter'
        pipe.metric_key,  # e.g. 'tweets'
        'avg',  # using the location key to tag the child
        columns=pipe.columns,  # if the columns are the same
        instance=pipe.instance_connector,  # optional: default is 'sql:main'
        tags=pipe.tags,  # optional: use the new tagging feature to group pipes together
    )
    ### Do something to get your derivative data
    foo_df = get_foo(pipe.get_backtrack_data(1440))
    return child_pipe.sync(foo_df, **kw)
```

Option B would be to register the child pipes outside of the plugin and define different sync functions for the different keys. The above example took way too long to type on mobile.

Also, v0.5.14 just introduced the ability to add tags to pipes: just add a list of strings under the key `tags` (like when you define columns, see `edit pipes` or as a parameter in the constructor}, and you can filter by tags with `--tags` or `-t`. I'm writing the documentation and will update the website soon.

I hope this gave some more insight into ways to design your plugins! Also check out the [package docs](https://docs.meerschaum.io/Pipe/index.html) for useful functions like `get_data()`.
